### ISSUE TITLE
[0.12.0.0] Waypoint Generator could use more work XD

### ISSUE THREAD

**Issue by Kaillera** [Regalis11/Barotrauma#4869]
- [x] I have searched the issue tracker to check if the issue has already been reported.

**Description**
Improvements on the waypoint generator generated this:

![Capture](https://user-images.githubusercontent.com/5086682/105525018-78760f00-5cae-11eb-8bb1-c5c96a55e225.PNG)

**Version**
0.12.0.0

**Additional information**
[Dandelion](https://steamcommunity.com/sharedfiles/filedetails/?id=2312353651) Workshop Sub for reference.

**Comment by itchyOwl** [comment]
It's a known issue, but it's easy to fix those manually. Some of the issues are probably due to raycasting and should be easy to fix, but others are because we remove points that are too close to the sub or to other waypoints. That issue might be a bit difficult to solve.

**Comment by itchyOwl** [comment]
More images of the same issue:
![image](https://user-images.githubusercontent.com/5370738/106136396-eb312f80-6171-11eb-89b9-40f0595afc9b.png)

Missing waypoints (Hemulen and Typhon2):
![image](https://user-images.githubusercontent.com/5370738/106136467-069c3a80-6172-11eb-88f0-55c5e09978f7.png)
![image](https://user-images.githubusercontent.com/5370738/106136663-44995e80-6172-11eb-9583-f2dcc5c53082.png)

Missing links (Kastrull):
![image](https://user-images.githubusercontent.com/5370738/106136596-30556180-6172-11eb-9cf8-b3958dcb6b62.png)

**Comment by itchyOwl** [comment]
Consider adding waypoints above platforms, see Regalis11/Barotrauma#4624.

**Comment by itchyOwl** [comment]
A missing link (also happens elsewhere):
![image.png](https://images.zenhubusercontent.com/5c93b12055cf186934fc0a03/e456337c-04cc-46ad-981c-639247fd68ae)

**Comment by itchyOwl** [comment]
Fixed most of the issues mentioned above.

The missing link in adminmodule_02_abandoned was due to a missing gap.

All the missing links inside the sub should be fixed.

The generator should now add waypoints over platforms, when needed.

I couldn't fix all the cases where we fail to connect the waypoints, but I managed to reduced the cases and also added better feedback about it when the generator fails.

The console error is now more descriptive. The ids of potentially invalid waypoints are displayed in red, and the waypoints that are at the end of the path (linked to only one node) are shown in yellow.

Changes applied in Regalis11/Barotrauma-development@17264a6.
